core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -827,159 +848,170 @@ public T next() {
                     nextChunkIter = hi != null ? nextChunk.descendingIter(hi, hiInclusive) : nextChunk.descendingIter();
                 }
             }
+            lastKeyBeforeChunkDelete = new AtomicReference<>(nextChunk);
+            nextChunk.signIterator(this);
+            setNextKey();
         }
 
-        /**
-         * sets nextChunk, nextChunkIter, next and nextValue to the next key
-         * if such a key doesn't exists then next is set to Chunk.NONE and nextValue is set to null
-         */
-        private void findNextKeyInRange() {
-            if (nextChunkIter == null) {
-                return;
-            }
+
+        private Chunk<K, V> getNextChunk(Chunk<K, V> current) {
             if (!isDescending) {
-                ascend();
+                return current.next.getReference();
             } else {
-                descend();
+                ByteBuffer serializedMinKey = current.minKey;
+                Map.Entry<Object, Chunk<K, V>> entry = skiplist.lowerEntry(serializedMinKey);
+                if (entry == null) {
+                    return null;
+                } else {
+                    return entry.getValue();
+                }
             }
         }
 
-        private void ascend() {
-            if (!findNextChunk()) {
-                return;
-            }
-            // now look for first key in range
-            next = nextChunkIter.next();
-            ByteBuffer key = getKey(next, nextChunk);
-
-            while (!inBounds(key)) {
-                if (tooHigh(key)) {
-                    // if we reached a key that is too high then there is no key in range
-                    next = Chunk.NONE;
-                    nextHandle = null;
-                    return;
-                }
-                if (!findNextChunk()) {
-                    return;
-                }
-                next = nextChunkIter.next();
-                key = getKey(next, nextChunk);
-            }
-            // set next value
-            Handle h = nextChunk.getHandle(next);
-            if (h != null) {
-                nextHandle = h;
+
+        private Chunk.ChunkIter getChunkIter(Chunk<K, V> current) {
+            if (!isDescending) {
+                return current.ascendingIter();
             } else {
-                nextHandle = null;
+                return current.descendingIter();
             }
         }
 
-        private boolean findNextChunk() {
+
+        /**
+         * sets nextChunk, nextChunkIter, next and nextValue to the next key
+         * if such a key doesn't exists then next is set to Chunk.NONE and nextValue is set to null
+         */
+        private void setNextKey() {
+
+            Chunk<K, V> currentChunk = nextChunk;
+
+
             while (!nextChunkIter.hasNext()) { // chunks can have only removed keys
-                nextChunk = nextChunk.next.getReference(); // try next chunk
+                nextChunk.signoutIterator(this);
```

Comment:
how can you be sure you already sign-in to the next chunk?

Reply:
There is no problem signing out even if your are not signing in.
This is exactly what will happen in this loop if we skip a few chunks